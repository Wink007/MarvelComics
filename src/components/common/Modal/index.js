/**
 *  Wrap <Modal> over the component you need to show in modal
 *  <Modal> component should have a single child <ModalWrapper>
 *  which handles modal closing and common styles
 * */

// Core
import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import { node } from 'prop-types';

export default class Modal extends Component {
  static propTypes = {
    children: node,
  };

  static defaultProps = {
    children: <div />,
  };

  state = {
    isInitialized: false,
  };

  componentDidMount() {
    this.root = document.getElementById('root');
    this.modalRoot = document.getElementById('modal-root');

    this.element = document.createElement('div');
    this.element.classList.add('modal');

    this.modalRoot.appendChild(this.element);
    this.root.style.position = 'fixed';

    this.handleInitializationEnd();
  }

  componentWillUnmount() {
    this.modalRoot.removeChild(this.element);
    this.root.style.position = 'static';
  }

  handleInitializationEnd = () => {
    this.setState(() => ({
      isInitialized: true,
    }));
  };

  render() {
    const { children } = this.props;
    const { isInitialized } = this.state;
    return (
      isInitialized
        ? ReactDOM.createPortal(children, this.element)
        : null
    );
  }
}
