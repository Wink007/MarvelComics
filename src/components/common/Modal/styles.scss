@import 'themes/variables/_variables-default.scss';

.ModalWrapper {
  display: flex;
  position: fixed;
  width: 100%;
  height: 100%;
  overflow-y: hidden;
  cursor: pointer;
  //background-color: $color-dark-3;

  &[data-background='true'] {
    background-color: #000;
  }
}
