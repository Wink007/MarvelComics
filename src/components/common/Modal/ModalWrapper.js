/**
 * Use this component to
 * - handle modal closing event
 * - stopPropagation on the div that covers { children }
 * */

// Core
import React, { Component } from 'react';
import { func, node, bool } from 'prop-types';

// Custom
import detectMobile from 'utils/ContextProviderHelpers/detectMobile';

// Styles
import './styles.scss';


class ModalWrapper extends Component {
  static propTypes = {
    children: node,
    onClick: func.isRequired,
    background: bool,
  };

  static defaultProps = {
    background: false,
    children: null,
  };

  stopPropagation = (event) => {
    event.stopPropagation();
  };

  render() {
    const { children, onClick, background } = this.props;
    const deviceDetect = detectMobile();
    const { isMobile } = deviceDetect;
    return (
      <div // eslint-disable-line
        className="ModalWrapper"
        onClick={onClick}
        data-background={background ? 'true' : ''}
        data-device={isMobile ? 'mobile' : 'desktop'}
      >
        <div // eslint-disable-line
          onClick={this.stopPropagation}
        >
          {children}
        </div>
      </div>
    );
  }
}

export default ModalWrapper;
