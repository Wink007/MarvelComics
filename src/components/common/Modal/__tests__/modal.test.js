import React from 'react';
import PageObject from 'utils/test/pageObject';
import Modal from '..';

const props = {
  children: <div />,
};

const page = (new PageObject(props)).setComponent(Modal, false);

describe('Check if the <Modal /> component rendered',
  () => {
    document.body.innerHTML = '<div id="root"></div><div id="modal-root"></div>';
    it('+++ render correctly component', () => {
      expect(page.getShallowRendered()).toMatchSnapshot();
    });
  });
