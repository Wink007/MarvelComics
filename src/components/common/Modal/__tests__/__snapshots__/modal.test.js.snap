// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if the <Modal /> component rendered +++ render correctly component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Modal>
    <div />
  </Modal>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "portal",
    "props": Object {
      "children": <div />,
      "containerInfo": <div
        class="modal"
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": "div",
    },
    "type": Symbol(react.portal),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "portal",
      "props": Object {
        "children": <div />,
        "containerInfo": <div
          class="modal"
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      "type": Symbol(react.portal),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
